ui/lib/core/addon/components/client-pagination.ts

```diff
@@ -0,0 +1,82 @@
+import { action } from '@ember/object';
+import { pluralize } from 'ember-inflector';
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import config from 'vault/config/environment';
+import MutableArray from '@ember/array/mutable';
+
+const { DEFAULT_PAGE_SIZE } = config.APP;
+
+const getSliceIdxForCurrentPage = (pageSize: number, currentPage: number) => {
+  const pageIndex = currentPage - 1;
+  const first = pageIndex * pageSize;
+  const last = first + pageSize;
+  return [first, last];
+};
+
+interface Args {
+  items: MutableArray<unknown>;
+  itemNoun?: string;
+  filterQuery?: string;
+}
+
+/**
+ * * @module ClientPagination
+ * `ClientPagination` components are used to render a list of records from the store.
+ * The list will be automatically paginated when >15 items in the list. If the list is empty
+ * an empty state will show with calculated title and message. To add empty state actions,
+ * wrap them in `<:emptyActions>` within the component.
+ *
+ * @example
+ * ```js
+ * <ClientPagination @itemNoun="role" @filterQuery={{this.filter}} @items={{filter-items this.model this.filter}} >
+ *   <:emptyActions>
+ *     <LinkTo @route="dashboard">This link renders an an empty state action</LinkTo>
+ *   </:emptyActions>
+ *   <:item as |role|>
+ *     <div data-test-list-item-role={{role.id}}>
+ *       I have access to the iterated role info: {{role.name}}
+ *     </div>
+ *   </:item>
+ * </ListView>
+ * ```
+ *
+ */
+export default class ClientPaginationComponent extends Component<Args> {
+  @tracked currentPage = 1;
+  @tracked pageSize = DEFAULT_PAGE_SIZE as number;
+  pageThreshold = DEFAULT_PAGE_SIZE as number;
+
+  get itemNoun() {
+    return this.args.itemNoun || 'item';
+  }
+
+  get emptyTitle() {
+    const items = pluralize(this.itemNoun);
+    return `No ${items} ${this.args.filterQuery ? 'found' : 'yet'}`;
+  }
+
+  get emptyMessage() {
+    const items = pluralize(this.itemNoun);
+    const { filterQuery } = this.args;
+    if (filterQuery) {
+      return `There are no ${items} matching matching "${filterQuery}"`;
+    }
+    return `Your ${items} will be listed here. Add your first ${this.itemNoun} to get started.`;
```

Comment:
I feel like this should handle the filter messaging but the empty state for no items existing should be more flexible 🤔 Basing this off of the bloated logic added to [options-for-backend](https://github.com/hashicorp/vault/blob/772ca6e1bb3fdec9e7908c8904fce1870564bb24/ui/lib/core/addon/helpers/options-for-backend.js#L97) to manage the different empty state messages I think avoiding a generated message would be ideal. 

Can we pass the empty state component entirely to the pagination component? Instead of just the empty state actions?

Reply:
We can, and I considered that route. The reason I didn't is because A) adding empty states will definitely get repetitive, especially since most of them will be the same. And B) for consistency between the empty states. 
Also, since this component doesn't handle filtering (that's handled in the parent) it seems weird to me that it would handle one empty state but not the other, so maybe we just pass in the empty state either way